research/picai/fl_nnunet/nnunet_client.py

```diff
@@ -639,25 +714,54 @@ def get_properties(self, config: Config) -> Dict[str, Scalar]:
         """
-        Checks the dataloaders type and if it is a MultiThreadedAugmenter or
-        NonDetMultiThreadedAugmenter calls the _finish method to ensure they
-        are properly shutdown
+        The nnunet dataloader/augmenter uses multiprocessing under the hood, so the
+        shutdown method terminates the child processes gracefully
 
         Args:
             dataloader (DataLoader): The dataloader to shutdown
             dl_name (Optional[str]): A string that identifies the dataloader
                 to shutdown. Used for logging purposes. Defaults to None
         """
         if dataloader is not None and isinstance(dataloader, nnUNetDataLoaderWrapper):
-            if isinstance(dataloader.nnunet_dataloader, (NonDetMultiThreadedAugmenter, MultiThreadedAugmenter)):
-                if dl_name is not None:
-                    log(INFO, f"\tShutting down nnunet dataloader: {dl_name}")
-                dataloader.nnunet_dataloader._finish()
+            log(INFO, f"\tShutting down nnunet dataloader: {dl_name}")
+            dataloader.shutdown()
+
+        del dataloader
 
     def shutdown(self) -> None:
-        # Not entirely sure if processes potentially opened by nnunet
-        # dataloaders were being ended so ensure that they are ended here
+        """
+        The nnunet dataloader/augmenter uses multiprocessing under the hood, so the
+        shutdown method terminates the child processes gracefully
+        """
         self.shutdown_dataloader(self.train_loader, "train_loader")
         self.shutdown_dataloader(self.val_loader, "val_loader")
         self.shutdown_dataloader(self.test_loader, "test_loader")
         return super().shutdown()
+
+    def update_before_train(self, current_server_round: int) -> None:
+        # Was getting OOM errors that could only be fixed by manually cleaning up RAM
+        # https://github.com/pytorch/pytorch/issues/95462
+        # The above issue seems to be the situation I was in.
+        gc.collect()  # Cleans up unused variables
+        # As the linked issue above points out, calling gc.freeze() greatly reduces the
+        # overhead of garbage collection. (from 1.5s to 0.005s)
+        if current_server_round == 2:
+            # Collect runs even faster if we freeze after the end of the first iteration
+            # Likely because a lot of variables are created in the first pass
+            # If we freeze before the first pass, gc.collect has to check all those
+            # variables
+            gc.freeze()
+
+    def transform_gradients(self, losses: TrainingLosses) -> None:
+        """
+        Apply the gradient clipping performed by the default nnunet trainer. This is
+        the default behaviour for nnunet 2.5.1
+        """
+        torch.nn.utils.clip_grad_norm_(self.model.parameters(), self.max_grad_norm)
+
+    def update_after_train(self, local_steps: int, loss_dict: Dict[str, float], config: Config) -> None:
```

Comment:
One thing I just thought about is that if we don't have full client participation, this will only be called on clients who are sampled in the final round. We could do an extra round at the end of the training with full participation, enforce no training happens and call the gc.unfreeze(). Or we could require full participation as a short term fix and work towards supporting client sampling. What do you think?

Reply:
Would moving this to client.shutdown resolve the issue?

Reply:
The only downside to that is if you don't have full client participation AND are doing something with the client afterwards before shutting it down. The only thing that I can think that you would be doing with the client afterwards would be finetuning and inference, in which case maybe its not terrible if the frozen objects stay frozen in memory since their probably needed in both anyways

Reply:
Yeah I think that should work! I think that is fine for now and we can make note of it to change if becomes a problem